crates/hash/src/address_impl.rs

```diff
@@ -33,8 +36,43 @@ impl Address for Contract {
     }
 }
 
+/// Hash the [`SolutionData`] in a manner that treats `state_mutations` like a set.
+///
+/// Hashing occurs as follows:
+///
+/// - predicate_to_solve.contract.0
+impl Address for SolutionData {
+    fn content_address(&self) -> ContentAddress {
+        let mut hasher = <sha2::Sha256 as Digest>::new();
+
+        // Include the predicate address.
+        hasher.update(self.predicate_to_solve.contract.0);
+        hasher.update(self.predicate_to_solve.predicate.0);
+
+        // Decision variables must be in a particular order, required by the predicate
+        // that is being solved. Thus, we hash them in their existing order.
+        hasher.update(bytes_from_word(self.decision_variables.len() as Word));
+        for value in &self.decision_variables {
+            crate::hash_len_then_words(value, &mut hasher);
+        }
+
+        // State mutations are a set. In order to ensure the same CA is produced regardless
+        // of the ordering of the state mutations, we hash the mutations in order of `Key`.
+        hasher.update(bytes_from_word(self.state_mutations.len() as Word));
+        let mut ixs: Vec<_> = (0..self.state_mutations.len()).collect();
+        ixs.sort_by_key(|&ix| &self.state_mutations[ix].key);
```

Comment:
Is this also correct? Do order of state mutations not matter? What about if we allow these to be read by the vm? It might be useful for a prorgam to reason about the order (first mutation is foo etc.) in the same way we do `decision variables`

Reply:
> It might be useful for a prorgam to reason about the order (first mutation is foo etc.)

I guess in the case where it's useful for the keys to be indexable, I'm imagining it would be OK to require that they're ordered by key?

Regardless, I think this is more of a design choice to be made on our part. I.e. sure, maybe there are possible use cases where a user-specified order is useful, but the trade-off will be that these are no longer a set.

The benefits of them being a set would be 1. we wouldn't need to track their order in order to reconstruct solutions and 2. it maybe opens up the possibility of merging sets of mutations, though I'm not sure how useful that would be.

Under the current design and behaviour I believe we treat them entirely as sets everywhere *besides* in content addressing.

> It might be useful for a prorgam to reason about the order (first mutation is foo etc.) in the same way we do decision variables

Decision variables are a bit different as they have no keys - you can kind of think of the indices as the user-specified keys that must match the indices expected by the predicate being solved. I don't think this is really relevant to state mutations, where the keys used are related to contract storage layout.

Reply:
If you order them by key then you have to prefix with some data that gives you the order. Maybe that's ok. I'm just not sure im seeing the benefits here though. We already have to pull Dec vars out in order and it's almost an identical process with mutations. I don't think it helps with merging except for dedupping but that is more of an app level concern.

By doing this we take away the possibility of using the order for meaning. If there are apps where passing in keys as args and mutations is wasteful and should only be passed in by mutations then we can't really do that. An example that comes to mind is token transfers that have the to, from and amount in the dec vars and the mutations for no real reason. Then we compare them to make sure they are the same but this could all be avoided by just using the mutation keys directly

Reply:
> I'm just not sure im seeing the benefits here though.

We would no longer need to store an extra `mutation_index` for every mutation entry in the node DB just to reconstruct solutions.

> By doing this we take away the possibility of using the order for meaning

Yeah, we should get to the bottom of whether these are a set or an ordered list.

> If there are apps where passing in keys as args and mutations is wasteful and should only be passed in by mutations then we can't really do that. An example that comes to mind is token transfers that have the to, from and amount in the dec vars and the mutations for no real reason. Then we compare them to make sure they are the same but this could all be avoided by just using the mutation keys directly

Would this not still be possible using the `Key` ordering either way? E.g. if we want some op that provides the keys, could we not just specify that keys are provided in order? That way we'd keep the set benefits but still make it possible to access them?